rfcs/20200505-transactional-fs.md

```diff
@@ -0,0 +1,421 @@
+provide transactional access and eventual consistency.
+
+## Motivation
+
+Current persistent storage implementation in Tensorflow relies on certain guarantees that
+existing local file systems provide. But in the age of big data, local filesystems are not always
+sufficient and/or prefetching the data to local system and then uploading after processing can be error-prone
+and harder to implement for end users. Direct access to persistent storage that provides different gurantees is desirable.
+Cloud storage solutions offered by Amazon, Google and others and databases can be examples of such persistent storages.
+Moreover even though local file systems provide certain atomic-like transactions, they are on file level.
+For the use cases like checkpointing, transactions are emulated through creating a temp directory, adding files to there and then
+renaming/moving temp directory to final location. Such operations would also benefit from enhancements proposed in this RFC.
+
+Transactions can also help with some filesystem access inconsistencies that can happen while reading and writing checkpoints. For example in while on thread reading files from a directory other may be modifying the underlying files. This could lead to reading an inconsistent or corrupt set of files by the reader. With transactions, each thread can have different transaction tokens and underlying file system can choose to postpone modification of files by redirecting them to a temporary location and then moving it in place when transactions end.
+
+## User Benefit
+
+With this extension proposal, users will have a more stable access to cloud storage systems as well as checkpointing redundancy can improve.
+
+## Design Proposal
+
+This RFC proposes to extend the [filesystem plugin rfc][filesystem_plugin] api with transaction markers. There can be different levels of transactions. First level can be global transactions, user starts a transaction scope. Any operations done within this scope will be grouped in this transaction. This is easiest to implement but have drawbacks such as only one transaction can be present at a time and different type of transactions may not always be reordered.
+Alternative to this approach is having multiple transaction scopes. User can create a transaction token and pass it to filesystem operations for plugin to differentiate among independent transactions. This token can be per file or per directory level granularity. Even though per file would give most flexibility, per directory level of transaction detail is most likely sufficient.
+
+Filesystem plugins may choose to ignore the transaction scopes or can delay the operations until the termination of transaction scope.
+
+### Extension to existing filesystem implementation
+
+Existing filesystem C++ api can easily be expanded by addition of three methods, an opaque structure and possibly a helper class to support transactions.
+
+```cpp
+struct TransactionToken{
```

Comment:
One alternative, less intrusive, API modification is instead of passing transaction tokens to all FS methods optionally we can have a variant of GetFileSystemForFile which instead is like StartTransactionForFile which returns a FileSystem which, when dereferenced, commits the transaction. Implicitly then all operations made on this file system are a part of the transaction until committed.

WDYT?

Reply:
@alextp +1 for non-intrusive APIs

Reply:
If I understood correctly, that is a quite brilliant idea. You are talking about binding the token to FS object, at which, all FS operations done on this objects belong to same transaction. But since most existing code uses Env class for file operations which forward it to respective FS, it would require more changes in the code base, where all code using Env for file system interaction needs to be changed to use FS directly. If this is acceptable, I liked your idea better.
In C api, things can get a bit messier though, since it would mean constructing all TF_*Ops structures and a lot of gymnastics to get it right unless we require C api still requires tokens as proposed in this RFC, but [C++ wrappers](https://github.com/tensorflow/community/blob/9c248b5603256c0641a05dd202adb2942c8af500/rfcs/20190506-filesystem-plugin-modular-tensorflow.md#incorporating-c-api-into-filesystem-classes) incorporates the logic you brought up. But that would mean that no direct C api interaction with file systems in framework code.

Reply:
That's a good point that the naive less intrusive approach I suggested won't work.

I don't have a rejoinder now, but I'll try to think of other non-intrusive or less intrusive approaches.

Reply:
Just a note: Core TF would never directly call the filesystem C API outside of the C++ wrapper.

Reply:
@alextp, considering what @mihaimaruseac said, your initial proposal may be better for most use cases if we agree moving users from Env* to FS*, What do you think?

Reply:
I don't follow, how so?

Reply:
Something like

``` cpp
// Replace Env::Default()->(filesystems methods with filesystem
// Env::Default()->CreateDir(dirname);
FileSystem* FS=Env::Default()->StartTransactionForURI(fname);
FS->CreateDir(dirname);
//.....
// Env::Default()->RenameFile(...)
FS->RenameFile(....);
//...
// Env::Default()->DeleteDir(..)
FS->DeleteDir(dirname);
FS->EndTransaction();
```
Since with @mihaimaruseac said that C layer is always accessed through wrapper, Same logic can be implemented with plugins. 
But then the API should be extended a little bit so that users can continue an existing transaction.

This does not necessarily reduce the total lines of code that need to be changed but simplifies token tracing.

Reply:
Note that the `fs` returned by `Env::GetFileSystemForFile` depends *only* on the scheme part of the URI. That is, there is exactly one `FileSystem` instance returned for all `s3://` paths.

I'm thinking of a design where the plugin should be the one having a mapping between files and associated transaction tokens and an API usage such as:

```cc
Token StartTransactionalOperation() {
  auto token = Env::Default()->GetTokenForURI(fname);
  Env::Default()->AddToTransaction(token, fname); // this could be optional
  Env::Default()->AddToTransaction(token, dirname);
  ... // add more paths to the same transaction

  // now operate on paths as before
  Env::Default()->CreateDir(dirname);
  WriteToFile(fname);
  ReadFromFile(fname1);
  ...

  // return token if we want t keep the transaction outside the scope of this function
  return token;
}

void EndTransactionalOperation(Token token) {
  Env::Default()->CreateDir(dirname2); // suppose dirname2 was not added to the transaction
  // this will succeed but not be transactional

  Env::Default()->AddToTransaction(token, dirname3); // here too dirname3 was not added
  Env::Default()->CopyRecursively(dirname, dirname3);

  token->EndTransaction(); 
  // at this point we either have dirname3 with the entire contents or not at all
  // at this point, cloud filesystems will upload dirname3
  // dirname2 is outside transaction even here
}
```

This has the advantage that no API needs to be changed, only new methods need to be added. Plus, these methods can be no-ops in a default implementation. This looks very unintrusive. Can this do all that we want to do?

Reply:
> Note that the fs returned by Env::GetFileSystemForFile depends only on the scheme part of the URI. That is, there is exactly one FileSystem instance returned for all s3:// paths.

What I was thinking is like
``` cpp
class WrappingFilesystem:public FileSystem{
WrappingFileSystem(Filesystem* fs, TransactionToken* t):fs_(fs),token_(t){}

Status CreateDir(dirname){
  return fs_->CreateDir(dirname,token_);
}

Status DeleteDir(dirname){
  return fs_->DeleteDir(dirname,token_);
}

// other Filesystem methods

EndTransaction(){
  fs_->EndTransaction(token)
}

Status GetTransactionTokenForFile(const string& fname,TransactionToken** token){
 if(fname.empty()){
   *token=token;
   return Status::OK()
 }else{
   return fs_->GetTransactionTokenForFile(fname,token);
 }
}
private:
  FileSystem* fs_;
  TransactionToken* token_;
}

class Env{
Status StartTransactionForURI(path,std::unique_ptr<FileSystem>* result ){
Filesystem * fs=nullptr;
auto status=GetFilesystemForFile(path,&fs);
 if(status.ok())){
   TransactionToken *token;
  status=fs->StartTransaction(path,&token);
 *result=std::make_unique<WrappingFileSystem>(fs,token);
 }
return status;
}
}
```
Your approach may solve the issue with OP based i/o if it is possible to deduce the file names in advance. Even though this likely increases amount of code change and would probably require multiple manifest implementations on Filesystems,  it may work.

But be aware, it may not be possible for all cases to delay permanent storage operations( such as cloud upload) until the end of transaction. FS implementation may choose to start the transfer earlier and remove the file if transaction fails or filter the file access out until transaction completes. As I mentioned in above comments, transactions can not guarantee atomic permanent storage operations across processes.

Reply:
The modular design was planning to declare `FileSystem` a final class to prevent filesystems from being declared in the non-modular way.

Even when the file patterns are not known in advance, we can register them to the transaction token as soon as they are known, just before operating with them.

Code changes to do the registration would be in the same places we would need to add transaction support anyway. However, there won't be additional changes to the filesystem API itself, except the need to serialize/deserialize the token on the filesystems which support it (all others being no-op)

Reply:
Yes, it may work. Only issue I am having is assume that a file is being written and read at the same time with a writer and a reader thread. With passing tokens or using FS, it may be possible to ensure that reader reads a valid file,  the file being overwritten, after the write operation or failure to read the file at all. So in a scenario like below
`` `cpp
// thread1 
Token* WriteTensors(){
// calculate or receive fname and dirname
  auto token = Env::Default()->GetTokenForURI(fname);
  Env::Default()->AddToTransaction(token, fname); // this could be optional
  Env::Default()->AddToTransaction(token, dirname);
  ... // add more paths to the same transaction
  // now operate on paths as before
  Env::Default()->CreateDir(dirname);
  WriteToFile(fname);
  // return token if we want t keep the transaction outside the scope of this function
  return token;
}
// thread 2
Token* ReadTensors(){
// calculate or receive  fname and dirname, identical to writer
  auto token = Env::Default()->GetTokenForURI(fname);
  Env::Default()->AddToTransaction(token, fname); // this could be optional
  Env::Default()->AddToTransaction(token, dirname);
  while(!Env::Default()->IsDirectory(dirname).ok() || 
               !Env::Default()->FileExists(fname.ok()){
       sleep(0.5); // or do something else
  }
  ReadFile(fname);
  return token;

}

```

What would be the behavior expected from FS? When user tries to access the same file, is it in reader token or writer token?

Reply:
We have 2 transactions in this case and they cannot be serialized. We can decide this to be undefined behavior (not specifying anything to happen in this case, advising users to not write this type of code), we can say that the first transaction that starts execution blocks the other or we can make an error to add two tokens for the same file.

Probably each filesystem can decide depending on filesystem internals

Reply:
If you pass tokens to operations you can do and it doesn't require serialization. Assume that you have a file that is being updated continuously at some point in time and you have a watcher thread that is polling file for updates. Not a good design but could happen in large workflows. 
 If you pass tokens explicitly, FS can ensure that reader is reading a vaild file, either by blocking or delaying overwrites. This is how checkpointing is implemented manually right?
reader polls checkpoint directory all the time, and writer writes it to a temp directory first and does a move. Since in most traditional persistent storage, move is just locked update of the FS table, it looks like an atomic move operation. If reader lists the directory or starts reading a file from that directory while temp directory is written by writer thread(s), it will get the contents of the previous iteration. Of course this is across processes but it could very well be a possible in-process case for large workflows.

Also I would rather be reluctant to leave the behavior to FS choice as you mentioned. This would lead to differences in shared models. These paths are usually given by the users and when sharing model it would behave differently. For example an TF example model taken from github would behave differently in local filesystems, cloud filesystems or other DB backed filesystems.

So this case should have a well defined behavior.

Reply:
We can block the operation to attach a file to a token if the file is already attached to the other token. This ensure blocking behavior and delaying overwrites.

Maybe this is actually the better implementation for concurrent transactions on the same file.

Reply:
In that case should filesystem ensure whether added object is directory or file? What happens in the case of directories. Also we have to make sure that appending to file uses same token, if done in another method. Which may require a unique filesystem based token key to be used when starting a transaction which would return existing token if there still an ongoing transaction with the same key or a new transaction if there isn't.

I am OK with this approach, but all these will start adding quite complicated layers of bookeeping into the filesystems which may have an impact on certain operations or filesystems.

Reply:
Do we prefer keeping API over complicating implementations?

Reply:
Adding the new parameter is intrusive and requires more changes than adding the declarative API to create a token and then attach objects to it. Modular filesystem design does not allow subclassing and only creates one single operations table per URI scheme, so core TF won't be able to pick between the subclass that has tokens and the one that ignores them.

Reply:
But if we relax subclassing of Modular filesystem or just make it something like base subsystem and let users always interact through wrapped filesystem, it won't be an issue right? Op table can contain the api with the tokens but wrapper may keep them as optional. There still will be one op table per URI and one filesystem. Users will access through filesystem pointer instead of Env::Default() and that would probably be only difference. This would be done purely on FW side and plugins only implement the api.